Enable `require-jsdoc` lint and add two lints related to React

somebody1234 opened the pull request:
### Pull Request Description
- Enables the `require-jsdoc` lint
- Fixes all lint errors caused by enabling this lint.

### Important Notes
- There is no option to require JSDoc for other constructs, like top-level constants.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

somebody1234 committed:
Add `require-jsdoc` lint, fix eslint errors, convert arrow functions to functions

somebody1234 committed:
Fix lints

somebody1234 commented:
oops :(

somebody1234 commented:
umm

somebody1234 commented:
not sure why prettier workflow failed, local run of `prettier -w .` doesn't change anything. it's says `lib/dashboard/src/authentication/src/authentication/cognito.ts` is formatted wrongly but i don't get any changes here

somebody1234 committed:
Bump prettier version; run prettier

PabloBuchu commented:
@somebody1234 even this PR has some weird issues on first electron open after build. Would you mind taking a look?

<img width="1728" alt="Screenshot 2023-04-26 at 09 05 28" src="https://user-images.githubusercontent.com/2855109/234496355-34bbaf15-9072-4d3d-8162-1ec635396165.png">

somebody1234 commented:
@PabloBuchu hmm... unfortunately still can't repro :(
steps taken:
```
./run ide build --skip-version-check --skip-wasm-opt --wasm-profile=dev --backend-source release --backend-rele
ase nightly
```
```
~/dist/ide/enso-linux-2023.1.1-dev.AppImage -debug.dev-tools -authentication -feature-preview.new-dashboard
```
the first time i wasn't logged in though, rebuilt and still cannot repro. i can try a clean build?

wdanilo commented on app/ide-desktop/lib/client/src/index.ts:
```diff
@@ -114,14 +116,11 @@ class App {
         }
     }
 
-    /** Set Chrome options based on the app configuration. For comprehensive list of available
+    /** Sets Chrome options based on the app configuration. For comprehensive list of available
```
In all of our Rust codebase we have the opposite rule - instead of "Sets" we use "Set". I think we should be consistent here. Using docs in the imperative form rather than 3rd person form makes them shorter and unified with Rust codebase (which is unified with official Rust style guide).

somebody1234 replied:
ah - i think the convention for TS in the other files was to use the opposite, but should be easy enough to change back

wdanilo commented on app/ide-desktop/lib/client/src/index.ts:
```diff
@@ -130,8 +129,10 @@ class App {
         }
-        const add = (option: string, value?: string) =>
+        /** Adds the specified Chrome option. */
+        function add(option: string, value?: string) {
             chromeOptions.push(new configParser.ChromeOption(option, value))
+        }
```
why we need to change const lambdas int functions? This makes the code more verbose.

wdanilo commented on app/ide-desktop/lib/client/src/project-management.ts:
```diff
@@ -66,13 +65,14 @@ export function importBundle(bundlePath: string): string {
+    // To be more resilient against different ways that user might attempt to create a bundle,
+    // we try to support both archives that:
+    // * contain a single directory with the project files - that directory name will be used
+    //   to generate a new target directory name;
+    // * contain the project files directly - in this case, the archive filename will be used
+    //   to generate a new target directory name.
+    // We try to tell apart these two cases
+    // by looking at the common prefix of the paths of the files in the archive.
     // If there is any, everything is under a single directory, and we need to strip it.
```
these docs are strangely formatted now. Some lines are too short.

wdanilo commented on app/ide-desktop/eslint.config.js:
```diff
@@ -226,6 +226,18 @@ export default [
+                        FunctionDeclaration: true,
+                        MethodDefinition: true,
+                        ClassDeclaration: true,
+                        ArrowFunctionExpression: true,
```
I think that requiring docs on arrow fn exprs is an overkill. Vast majority of them are small helper functions that do not require better docs than just a good name. In case we will find a code during review with too long such an arrow expression, we can always refactor it to a normal fn :)

wdanilo commented on app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/userMenu.tsx:
```diff
@@ -16,6 +16,7 @@ interface UserMenuItemProps {
     onClick?: React.MouseEventHandler<HTMLDivElement>
 }
 
+/** User menu item */
```
missing dot

wdanilo reviewed (approved):
Oh wow, I feel for you. Looks like a pretty tiring job with adding docs everywhere!

wdanilo commented:
Ofc, please remember about Q/A before merging.

somebody1234 committed:
Address review

somebody1234 committed:
Merge branch 'develop' into wip/sb/require-jsdoc

somebody1234 committed:
Fix lint errors; increase coverage of `require-jsdoc` lint; add more lints

somebody1234 committed:
Add sections

somebody1234 commented:
files with missing sections found using this fish shell command:
```fish
for f in (find . -not \( -name node_modules -prune \) \( -name '*.ts' -o -name '*.tsx' -o -name '*.js' \) ); grep "====" $f > /dev/null || echo $f; end
```

somebody1234 committed:
Merge branch 'develop' into wip/sb/require-jsdoc

somebody1234 committed:
Get rid of newlines adjacent to `/**` and `*/`

somebody1234 committed:
Make all non-JSX lines under 100 characters long

somebody1234 commented:
`/**` and `*/` being on their own line should now be all removed.
the two exceptions are after `@see` which incorrectly triggers the `jsdoc/no-multi-asterisks` lint.
```
grep -Ir '/\*\*$\|^ \+\*/' --exclude-dir=node_modules
```

similarly with all lines over 100 characters long
```
grep -Ir '.\{101,\}' --exclude-dir=node_modules --include='*.ts' --include='*.js'
```

somebody1234 committed:
Merge branch 'develop' into wip/sb/require-jsdoc

somebody1234 committed:
Fix lints

wdanilo commented:
I was reviewing this a week ago - still no Q/A - who is handling Q/A here?

somebody1234 committed:
Merge branch 'develop' into wip/sb/require-jsdoc

somebody1234 commented:
@PabloBuchu @indiv0 could one of you perhaps do a QA?

somebody1234 committed:
Merge branch 'develop' into wip/sb/require-jsdoc

mwu-tow commented:
@somebody1234 
Please update this against the current develop. I've tried building this and running but I see some issues in the node editor that are not present on develop. I suppose they are not related to your changes, but still, I'd like to be able to QA this after merge conflicts are resolved.

somebody1234 committed:
Merge branch 'develop' into wip/sb/require-jsdoc

somebody1234 commented:
@mwu-tow merged, but i'm not sure any of the commits merged in would fix the issue you encountered - it was only missing the last 2 commits, so:
- the TS lint revert
- the "fix engine version check" commit

mwu-tow commented:
@somebody1234  
Huh, to be honest, I have no idea why (it might have been just rebuilt) but after testing now all looks good!
🟢 and sorry for the confusion.

PabloBuchu commented:
@somebody1234 please merge with develop and solve conflicts so we can merge it :)

somebody1234 committed:
Merge branch 'develop' into wip/sb/require-jsdoc

PabloBuchu commented:
@somebody1234 ping

somebody1234 committed:
Merge branch 'develop' into wip/sb/require-jsdoc

somebody1234 committed:
Fix lint errors and comment formatting

somebody1234 commented:
whoops, fixed @PabloBuchu. there are some minor changes, but they should all be changes to comments

sylwiabr commented:
@somebody1234 what is the status here?

somebody1234 commented:
@sylwiabr should be ready, it's just had quite a few merge conflicts

somebody1234 committed:
Merge branch 'develop' into wip/sb/require-jsdoc

PabloBuchu commented:
QA 🟢

somebody1234 committed:
Merge branch 'develop' into wip/sb/require-jsdoc

somebody1234 committed:
Fix lint errors; minor refactor

somebody1234 commented:
QA:
dashboard actions, tested using `watch-dashboard`.
:heavy_check_mark: rename project
:heavy_check_mark: delete project
:heavy_check_mark: create empty project
:heavy_check_mark: create project from template
:heavy_check_mark: delete project
:heavy_check_mark: create folder
:heavy_check_mark: create secret
:heavy_check_mark: delete secret
:heavy_check_mark: create file (upload via form) - note: errors when the file does not have an extension with "unable to upload file". it may be desirable to pass-through the error "file name must be in the format <file_name>.<extension>" instead.
:heavy_check_mark: create file (button at top)
:heavy_check_mark: delete file
:heavy_check_mark: open help chat
:heavy_check_mark: filter using search bar
:heavy_check_mark: open user menu
:heavy_check_mark: change password (user menu) - `onSubmit` is fired twice; fixed in another PR.
:heavy_check_mark: sign out (user menu)
tested using `ide build`:
:heavy_check_mark: log in (email + password)
:heavy_check_mark: forgot password
:heavy_check_mark: reset password (incorrect code)
:heavy_check_mark: reset password (correct code)
:heavy_check_mark: switch backends
:warning: oauth login works - i'm on linux so i'm unable to test. however, i got this warning while attempting to login anyway:
```
VM4 sandbox_bundle:2 MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 open-deep-link listeners added
```
testing that frontend recompile watcher works:
:heavy_check_mark: `ide watch`
testing that live reload works:
:heavy_check_mark: `gui watch` - was navigating to `/` on reload, losing the query string; fixed in another PR
:heavy_check_mark: `watch-dashboard`

flags, tested using `ide build`:
:heavy_check_mark: `-startup.entry=grid_view`, dashboard enabled (should not appear on alternate entry points)
:heavy_check_mark: `-startup.entry=grid_view`
:warning: `-startup.project`, dashboard enabled - this is being fixed in another PR.
:heavy_check_mark: `-startup.project`

somebody1234 committed:
Minor fixes
